File: src/main/java/com/anod/appwatcher/utils/AppLog.java
package com.anod.appwatcher.utils;

import android.util.Log;

import com.android.volley.VolleyError;
import com.anod.appwatcher.BuildConfig;
import com.crashlytics.android.Crashlytics;

import java.util.IllegalFormatException;
import java.util.Locale;

public class AppLog {
	private final static String TAG = "AppWatcher";

    public static final boolean DEBUG;

    static {
        DEBUG = Log.isLoggable("AppWatcher", BuildConfig.DEBUG ? Log.DEBUG : Log.WARN);
    }

	public static void d(String msg) {
		if (DEBUG) Log.d(TAG, format(msg));
	}

    public static void d(final String msg, final Object... params) {
        if (DEBUG) Log.d(TAG, format(msg, params));
    }

	public static void v(String msg) {
		Log.v(TAG, format(msg));
	}
	
	public static void e(String msg) {
        Log.e(TAG, format(msg));
	}

	public static void e(String msg, Throwable tr) {
        Log.e(TAG, format(msg), tr);
        Crashlytics.logException(tr);
	}

    public static void e(String msg, final Object... params) {
        Log.e(TAG, format(msg, params));
    }

    public static void e(VolleyError error) {
         e("Volley: "+error.getClass().getSimpleName()+": "+error.getMessage(),error.getCause());
    }

    public static void ex(Throwable e) {
        Log.e(TAG, e.getMessage(), e);
        Crashlytics.logException(e);
    }

    public static void w(String msg) {
        Log.v(TAG, format(msg));
    }

    public static void v(String msg, final Object... params) {
        Log.v(TAG, format(msg, params));
    }

    private static String format(final String msg, final Object... array) {
        String formatted;
        if (array == null || array.length == 0) {
            formatted = msg;
        } else {
            try {
                formatted = String.format(Locale.US, msg, array);
            }
            catch (IllegalFormatException ex) {
                e("IllegalFormatException: formatString='%s' numArgs=%d", msg, array.length);
                formatted = msg + " (An error occurred while formatting the message.)";
            }
        }
        final StackTraceElement[] stackTrace = new Throwable().fillInStackTrace().getStackTrace();
        String string = "<unknown>";
        for (int i = 2; i < stackTrace.length; ++i) {
            final String className = stackTrace[i].getClassName();
            if (!className.equals(AppLog.class.getName())) {
                final String substring = className.substring(1 + className.lastIndexOf(46));
                string = substring.substring(1 + substring.lastIndexOf(36)) + "." + stackTrace[i].getMethodName();
                break;
            }
        }
        return String.format(Locale.US, "[%d] %s: %s", Thread.currentThread().getId(), string, formatted);
    }


}

File: framework/src/main/kotlin/info/anodsplace/framework/AppLog.kt
package info.anodsplace.framework

import android.os.Looper
import android.util.Log
import java.util.*

/**
 * @author alex
 * @date 2015-05-11
 */
class AppLog {
    var listener: Listener? = null

    interface Logger {
        fun println(priority: Int, tag: String, msg: String)

        open class Android : Logger {
            override fun println(priority: Int, tag: String, msg: String) {
                Log.println(priority, tag, msg)
            }
        }

        open class StdOut : Logger {
            override fun println(priority: Int, tag: String, msg: String) {
                println("[$tag:$priority] $msg")
            }
        }
    }

    interface Listener {
        fun onLogException(tr: Throwable)
    }

    companion object {

        var tag = "AppLog"
        var level = Log.INFO
        val instance: AppLog by lazy { AppLog() }

        val isVerbose = level <= Log.VERBOSE
        val isDebug = level <= Log.DEBUG

        var logger: Logger = Logger.Android()

        fun setDebug(buildConfigDebug: Boolean, loggableTag: String) {
            val isDebug = buildConfigDebug || Log.isLoggable(loggableTag, Log.DEBUG)
            level = if (isDebug) {
                Log.DEBUG
            } else {
                Log.INFO
            }
        }

        fun d(msg: String) {
            log(Log.DEBUG, format(msg, null))
        }

        fun i(msg: String, tag: String?) {
            log(Log.INFO, format(msg, tag))
        }

        fun v(msg: String) {
            log(Log.VERBOSE, format(msg, null))
        }

        fun e(msg: String, tag: String?) {
            loge(format(msg, tag), null)
        }

        fun e(msg: String, tag: String, tr: Throwable) {
            loge(format(msg, tag), tr)
            instance.listener?.onLogException(tr)
        }

        fun e(msg: String, tr: Throwable) {
            loge(format(msg, null), tr)
            instance.listener?.onLogException(tr)
        }

        fun e(tr: Throwable) {
            val message = tr.message ?: "Throwable is null"
            e(message, tr)
        }

        fun e(msg: String, tag: String, vararg params: Any) {
            loge(format(msg, tag, *params), null)
        }

        fun e(msg: String, vararg params: Any) {
            loge(format(msg, null, *params), null)
        }

        fun w(msg: String, tag: String?) {
            log(Log.VERBOSE, format(msg, tag))
        }

        fun v(msg: String, vararg params: Any) {
            log(Log.VERBOSE, format(msg, null, *params))
        }

        private fun log(priority: Int, msg: String) {
            if (priority >= level) {
                logger.println(priority, tag, msg)
            }
        }

        private fun loge(message: String, tr: Throwable?) {
            val trace = if (logger is Logger.Android) Log.getStackTraceString(tr) else ""
            logger.println(Log.ERROR, tag, message + '\n'.toString() + trace)
        }

        private fun format(msg: String, method: String?, vararg array: Any): String {
            val formatted: String = if (array.isEmpty()) {
                msg
            } else {
                try {
                    String.format(Locale.US, msg, *array)
                } catch (ex: IllegalFormatException) {
                    e("IllegalFormatException: formatString='%s' numArgs=%d", msg, array.size)
                    "$msg (An error occurred while formatting the message.)"
                }
            }
            val stackTrace = Throwable().fillInStackTrace().stackTrace
            val tag = if (method == null) {
                var methodFromTrace = "<unknown>"
                for (i in 2 until stackTrace.size) {
                    val className = stackTrace[i].className
                    if (className != AppLog::class.java.name) {
                        val substring = className.substring(1 + className.lastIndexOf(46.toChar()))
                        methodFromTrace = substring.substring(1 + substring.lastIndexOf(36.toChar())) + "." + stackTrace[i].methodName
                        break
                    }
                }
                methodFromTrace
            } else {
                method
            }
            val isMain = Looper.myLooper() == Looper.getMainLooper()
            return String.format(Locale.US, "[%s%d] %s: %s",
                    if (isMain) "MAIN:" else "",
                    Thread.currentThread().id,
                    tag,
                    formatted)
        }
    }
}
